\documentclass[11pt,]{article}
\usepackage[left=1in,top=1in,right=1in,bottom=1in]{geometry}
\newcommand*{\authorfont}{\fontfamily{phv}\selectfont}
\usepackage[]{mathpazo}


  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}



\usepackage{abstract}
\renewcommand{\abstractname}{}    % clear the title
\renewcommand{\absnamepos}{empty} % originally center

\renewenvironment{abstract}
 {{%
    \setlength{\leftmargin}{0mm}
    \setlength{\rightmargin}{\leftmargin}%
  }%
  \relax}
 {\endlist}

\makeatletter
\def\@maketitle{%
  \newpage
%  \null
%  \vskip 2em%
%  \begin{center}%
  \let \footnote \thanks
    {\fontsize{18}{20}\selectfont\raggedright  \setlength{\parindent}{0pt} \@title \par}%
}
%\fi
\makeatother




\setcounter{secnumdepth}{0}

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}


\title{Composite Kernel Machine Regression based on Likelihood Ratio Test with
Application for Combined Genetic and Gene-environment Interaction Effect  }



\author{\Large Ni Zhao\vspace{0.05in} \newline\normalsize\emph{Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA}   \and \Large Haoyu Zhang\vspace{0.05in} \newline\normalsize\emph{Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA}   \and \Large Jennifer J. Clark\vspace{0.05in} \newline\normalsize\emph{Food and Drug Administration, Baltimore, MD, USA}   \and \Large Arnab Maity\vspace{0.05in} \newline\normalsize\emph{Department of Statistics, North Carolina State University, Raleigh,
NC,USA}   \and \Large Michael C. Wu\vspace{0.05in} \newline\normalsize\emph{Public Health Sciences Division, Fred Hutchinson Cancer Research Center,
Seattle, WA, USA}  }


\date{}

\usepackage{titlesec}

\titleformat*{\section}{\normalsize\bfseries}
\titleformat*{\subsection}{\normalsize\itshape}
\titleformat*{\subsubsection}{\normalsize\itshape}
\titleformat*{\paragraph}{\normalsize\itshape}
\titleformat*{\subparagraph}{\normalsize\itshape}


\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage[strings]{underscore} % protect underscores in most circumstances



\newtheorem{hypothesis}{Hypothesis}
\usepackage{setspace}

\makeatletter
\@ifpackageloaded{hyperref}{}{%
\ifxetex
  \PassOptionsToPackage{hyphens}{url}\usepackage[setpagesize=false, % page size defined by xetex
              unicode=false, % unicode breaks when used with xetex
              xetex]{hyperref}
\else
  \PassOptionsToPackage{hyphens}{url}\usepackage[unicode=true]{hyperref}
\fi
}

\@ifpackageloaded{color}{
    \PassOptionsToPackage{usenames,dvipsnames}{color}
}{%
    \usepackage[usenames,dvipsnames]{color}
}
\makeatother
\hypersetup{breaklinks=true,
            bookmarks=true,
            pdfauthor={Ni Zhao (Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA) and Haoyu Zhang (Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA) and Jennifer J. Clark (Food and Drug Administration, Baltimore, MD, USA) and Arnab Maity (Department of Statistics, North Carolina State University, Raleigh,
NC,USA) and Michael C. Wu (Public Health Sciences Division, Fred Hutchinson Cancer Research Center,
Seattle, WA, USA)},
             pdfkeywords = {gene-environment interactions, kernel machine testing, likelihood ratio
test, multiple variance compo- nents, spectral decomposition,
unidentifiable conditions},  
            pdftitle={Composite Kernel Machine Regression based on Likelihood Ratio Test with
Application for Combined Genetic and Gene-environment Interaction Effect},
            colorlinks=true,
            citecolor=blue,
            urlcolor=blue,
            linkcolor=magenta,
            pdfborder={0 0 0}}
\urlstyle{same}  % don't use monospace font for urls

% set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother



% add tightlist ----------
\providecommand{\tightlist}{%
\setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}

\begin{document}
	
% \pagenumbering{arabic}% resets `page` counter to 1 
%
% \maketitle

{% \usefont{T1}{pnc}{m}{n}
\setlength{\parindent}{0pt}
\thispagestyle{plain}
{\fontsize{18}{20}\selectfont\raggedright 
\maketitle  % title \par  

}

{
   \vskip 13.5pt\relax \normalsize\fontsize{11}{12} 
\textbf{\authorfont Ni Zhao} \hskip 15pt \emph{\small Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA}   \par \textbf{\authorfont Haoyu Zhang} \hskip 15pt \emph{\small Department of Biostatistics, Johns Hopkins University, Baltimore, MD,
USA}   \par \textbf{\authorfont Jennifer J. Clark} \hskip 15pt \emph{\small Food and Drug Administration, Baltimore, MD, USA}   \par \textbf{\authorfont Arnab Maity} \hskip 15pt \emph{\small Department of Statistics, North Carolina State University, Raleigh,
NC,USA}   \par \textbf{\authorfont Michael C. Wu} \hskip 15pt \emph{\small Public Health Sciences Division, Fred Hutchinson Cancer Research Center,
Seattle, WA, USA}   

}

}








\begin{abstract}

    \hbox{\vrule height .2pt width 39.14pc}

    \vskip 8.5pt % \small 

\noindent Most common human diseases are a result from the combined effect of
genes, the environmental factors and their interactions such that
including gene-environment (GE) interactions can improve power in gene
mapping studies. The standard strategy is to test the SNPs, one-by-one,
using a regression model that includes both the SNP effect and the GE
interaction. However, the SNP-by-SNP approach has serious limitations,
such as the inability to model epistatic SNP effects, biased estimation
and reduced power. Thus, in this paper, we develop a kernel machine
regression framework to model the overall genetic effect of a SNP-set,
considering the possible GE interaction. Specifically, we use a
composite kernel to specify the overall genetic effect via a
nonparametric function and we model additional covariates parametrically
within the regression framework. The composite kernel is constructed as
a weighted average of two kernels, one corresponding to the genetic main
effect and one corresponding to the GE interaction effect. We propose a
likelihood ratio test (LRT) and a restricted likelihood ratio test
(RLRT) for statistical significance. We derive a Monte Carlo approach
for the finite sample distributions of LRT and RLRT statistics.
Extensive simulations and real data analysis show that our proposed
method has correct type I error and can have higher power than
score-based approaches under many situations.


\vskip 8.5pt \noindent \emph{Keywords}: gene-environment interactions, kernel machine testing, likelihood ratio
test, multiple variance compo- nents, spectral decomposition,
unidentifiable conditions \par

    \hbox{\vrule height .2pt width 39.14pc}



\end{abstract}


\vskip 6.5pt


\noindent  \section{Overview}\label{overview}

This vegnette provides an introduction to the `CKLRT' package. To load
the package, users need to install package from CRAN and CKLRT from
github. The package can be loaded with the following command:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#install.packages("devtools")  }
\KeywordTok{library}\NormalTok{(devtools)  }
\CommentTok{#install_github("andrewhaoyu/CKLRT")}
\KeywordTok{library}\NormalTok{(CKLRT)}
\end{Highlighting}
\end{Shaded}

\section{Example}\label{example}

In this vegnette, we will decomstrate the methods with a simple example.
1. X present other covariates we want to adjust. 2. E represents the
environment variable. 3. G is is the genotype matrix with two SNPs
inside 4. y is the simulated outcomes.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(mgcv); }\KeywordTok{library}\NormalTok{(MASS); }\KeywordTok{library}\NormalTok{(nlme);}
\KeywordTok{library}\NormalTok{(compiler);}\KeywordTok{library}\NormalTok{(Rcpp);}\KeywordTok{library}\NormalTok{(RcppEigen)}
\KeywordTok{library}\NormalTok{(CKLRT)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{6}\NormalTok{)}
\NormalTok{n =}\StringTok{ }\DecValTok{200} \CommentTok{# the number of observations}
\NormalTok{X =}\StringTok{ }\KeywordTok{rnorm}\NormalTok{(n) }\CommentTok{# the other covariates}
\NormalTok{p =}\StringTok{ }\DecValTok{2} \CommentTok{# two snp in a gene will be simulated}
\NormalTok{G =}\StringTok{ }\KeywordTok{runif}\NormalTok{(n}\OperatorTok{*}\NormalTok{p)}\OperatorTok{<}\StringTok{ }\FloatTok{0.5}
\NormalTok{G =}\StringTok{ }\NormalTok{G }\OperatorTok{+}\StringTok{ }\KeywordTok{runif}\NormalTok{(n}\OperatorTok{*}\NormalTok{p) }\OperatorTok{<}\StringTok{ }\FloatTok{0.5}
\NormalTok{G =}\StringTok{ }\KeywordTok{matrix}\NormalTok{(G, n,p) }\CommentTok{#genetic matrix}
\NormalTok{E =}\StringTok{ }\NormalTok{(}\KeywordTok{runif}\NormalTok{(n) }\OperatorTok{<}\StringTok{ }\FloatTok{0.5}\NormalTok{)}\OperatorTok{^}\DecValTok{2} \CommentTok{#enviroment effect}
\NormalTok{y =}\StringTok{ }\KeywordTok{rnorm}\NormalTok{(n) }\OperatorTok{+}\StringTok{ }\NormalTok{G[,}\DecValTok{1}\NormalTok{] }\OperatorTok{*}\StringTok{ }\FloatTok{0.3} \CommentTok{#observations}
\CommentTok{#apply the likelihood ratio test}
\KeywordTok{omniLRT_fast}\NormalTok{(y, }\DataTypeTok{X =}  \KeywordTok{cbind}\NormalTok{(X, E),}\DataTypeTok{K1 =}\NormalTok{ G }\OperatorTok{%*%}\StringTok{ }\KeywordTok{t}\NormalTok{(G),}\DataTypeTok{K2 =}\NormalTok{ (G}\OperatorTok{*}\NormalTok{E) }\OperatorTok{%*%}\StringTok{ }\KeywordTok{t}\NormalTok{(G }\OperatorTok{*}\StringTok{ }\NormalTok{E))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $p.dir
## [1] 0.0248
## 
## $p.aud
## [1] 0.02538226
## 
## $LR
## [1] 3.408
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#apply the restricted likelihood ratio test}
\KeywordTok{omniRLRT_fast}\NormalTok{(y, }\DataTypeTok{X =}  \KeywordTok{cbind}\NormalTok{(X, E),}\DataTypeTok{K1 =}\NormalTok{ G }\OperatorTok{%*%}\StringTok{ }\KeywordTok{t}\NormalTok{(G),}\DataTypeTok{K2 =}\NormalTok{ (G}\OperatorTok{*}\NormalTok{E) }\OperatorTok{%*%}\StringTok{ }\KeywordTok{t}\NormalTok{(G }\OperatorTok{*}\StringTok{ }\NormalTok{E))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $p.dir
## [1] 0.0228
## 
## $p.aud
## [1] 0.01961945
## 
## $LR
## [1] 3.701362
\end{verbatim}

The results of the function contain three elements: 1. p.dir is the
p-value of likelihood ratio test based on emprical distrition. 2. p.aud
is the p-value by approximating the null distribution as a mixture of a
point mass at zero with probability b and weighted chi square
distribution with d degrees of freedom with probality of 1-b. 3. LR is
the likelihood ratio test statistics.

\section{References}\label{references}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  N. Zhao, H. Zhang, J. Clark, A. Maity, M. Wu. Composite Kernel Machine
  Regression based on Likelihood Ratio Test with Application for
  Combined Genetic and Gene-environment Interaction Effect (Submitted)
\end{enumerate}




\newpage
\singlespacing 
\end{document}